runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleExecutionState.java

```diff
@@ -149,11 +149,40 @@ public String getLullMessage(Thread trackedThread, Duration millis) {
     return message.toString();
   }
 
+  @VisibleForTesting
+  public String getBundleLullMessage(Thread trackedThread, Duration millis) {
+    // TODO(ajamato): Share getBunldeLullMessage code with DataflowExecutionState.
+    String userStepName =
+        this.labelsMetadata.getOrDefault(MonitoringInfoConstants.Labels.PTRANSFORM, null);
+    StringBuilder message = new StringBuilder();
+    message.append("Operation ongoing");
+    if (userStepName != null) {
+      message.append(" in bundle ");
+    }
+    message
+        .append(" for at least ")
+        .append(formatDuration(millis))
+        .append(" without outputting or completing ")
+        .append(getStateName());
+    message.append("\n");
+
+    StackTraceElement[] fullTrace = trackedThread.getStackTrace();
+    for (StackTraceElement e : fullTrace) {
+      message.append("  at ").append(e).append("\n");
+    }
+    return message.toString();
+  }
+
   @Override
   public void reportLull(Thread trackedThread, long millis) {
     LOG.warn(getLullMessage(trackedThread, Duration.millis(millis)));
   }
 
+  @Override
+  public void reportBundleLull(Thread trackedThread, String customLogMessage, long millis) {
```

Comment:
if this is a method on ExecutionStateTracker, you could get rid of the customLogMessage by having DataflowExecutionStateTracker override log it's messaeg and then call the super method.

Reply:
This would trigger a bigger change because the logging is happening at DataflowExecutionState which is an implementation of ExecutionState.

Reply:
see above, seems worthwhile changing even if it makes PR bigger.